pennylane_lightning/core/src/simulators/lightning_gpu/algorithms/AdjointJacobianGPU.hpp

```diff
@@ -0,0 +1,446 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc. and contributors.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file AdjointJacobianGPU.hpp
+ */
+
+#pragma once
+
+#include <chrono>
+#include <future>
+#include <omp.h>
+#include <span>
+#include <thread>
+#include <variant>
+
+#include "AdjointJacobianBase.hpp"
+#include "DevTag.hpp"
+#include "DevicePool.hpp"
+#include "LinearAlg.hpp"
+#include "ObservablesGPU.hpp"
+
+/// @cond DEV
+namespace {
+using namespace Pennylane::Algorithms;
+using namespace Pennylane::LightningGPU;
+using namespace Pennylane::LightningGPU::Observables;
+namespace cuUtil = Pennylane::LightningGPU::Util;
+} // namespace
+/// @endcond
+
+namespace Pennylane::LightningGPU::Algorithms {
+
+/**
+ * @brief GPU-enabled adjoint Jacobian evaluator following the method of
+ * arXiV:2009.02823
+ *
+ * @tparam StateVectorT State vector type.
+ */
+template <class StateVectorT>
+class AdjointJacobian final
+    : public AdjointJacobianBase<StateVectorT, AdjointJacobian<StateVectorT>> {
+  private:
+    using ComplexT = typename StateVectorT::ComplexT;
+    using PrecisionT = typename StateVectorT::PrecisionT;
+    using CFP_t = decltype(cuUtil::getCudaType(PrecisionT{}));
```

Comment:
More CFP_t here and below

Reply:
As discussed, let's work on `CFP_t` in a follow-up PR later.